Panic using Extend in Type declaration

Using a design that has both extended attributes and an attribute of type MapOf(String, Any), which is also referenced in Body, I get a panic:
```go
package design

import (
	. "goa.design/goa/dsl"
)

var _ = API("Something", func() {
})

var CommonAttributes = Type("Common", func() {
	Token("bar", String)
})

var _ = Service("Foo", func() {
	Error("bad_request")

	HTTP(func() {
		Path("/foo")
		Response("bad_request", StatusBadRequest)
	})

	Method("create", func() {
		Payload(MyPayload)
		Result(func() {
			Attribute("location", String)
			Required("location")
		})

		HTTP(func() {
			POST("/")
			Body("fields")
			Response(StatusCreated, func() {
				Header("location:Location")
			})
		})
	})
})

var MyPayload = Type("MyPayload", func() {
	Extend(CommonAttributes)   // <--- seems to cause the issue somehow
	Attribute("fields", MapOf(String, Any))
	Required("bar", "fields")
})
```
```
$ goa gen foo/design
exit status 2
panic: cannot merge non object attributes

goroutine 1 [running]:
goa.design/goa/expr.(*AttributeExpr).Merge(0xc0002b0460, 0xc0000bcf00)
	/home/me/gopath/src/goa.design/goa/expr/attribute.go:291 +0x25e
goa.design/goa/expr.(*AttributeExpr).Finalize(0xc0002b0460)
	/home/me/gopath/src/goa.design/goa/expr/attribute.go:275 +0x176
goa.design/goa/expr.(*HTTPEndpointExpr).Finalize(0xc0002b03c0)
	/home/me/gopath/src/goa.design/goa/expr/http_endpoint.go:519 +0xb8a
goa.design/goa/eval.finalizeSet(0xc00005fa30, 0x1, 0x1, 0x0, 0x0)
	/home/me/gopath/src/goa.design/goa/eval/eval.go:124 +0xb1
goa.design/goa/expr.(*RootExpr).WalkSets(0xe99b80, 0xa57ee8)
	/home/me/gopath/src/goa.design/goa/expr/root.go:130 +0x9d2
goa.design/goa/eval.RunDSL(0x7ffc7476c9e9, 0xc000026520)
	/home/me/gopath/src/goa.design/goa/eval/eval.go:55 +0x423
main.main()
	/home/me/gopath/src/foo/goa070681146/main.go:47 +0x208
```
but then if I replace the extended attributes with a literal copy of the attributes, it works:
```
var MyPayload = Type("MyPayload", func() {
	Token("bar", String)
	Attribute("fields", MapOf(String, Any))
	Required("bar", "fields")
})
```
```
$ goa gen foo/design
gen/foo/client.go
gen/foo/endpoints.go
gen/foo/service.go
gen/http/cli/something/cli.go
gen/http/foo/client/cli.go
gen/http/foo/client/client.go
gen/http/foo/client/encode_decode.go
gen/http/foo/client/paths.go
gen/http/foo/client/types.go
gen/http/foo/server/encode_decode.go
gen/http/foo/server/paths.go
gen/http/foo/server/server.go
gen/http/foo/server/types.go
gen/http/openapi.json
gen/http/openapi.yaml
```

--- a/dsl/http.go
+++ b/dsl/http.go
@@ -699,16 +699,11 @@ func Body(args ...interface{}) {
 	)
 	switch a := args[0].(type) {
 	case string:
-		if ref.Find(a) == nil {
-			eval.ReportError("%q is not found in result type", a)
-			return
-		}
-		obj := expr.AsObject(ref.Type)
-		if obj == nil {
+		if !expr.IsObject(ref.Type) {
 			eval.ReportError("%s type must be an object with an attribute with name %#v, got %T", kind, a, ref.Type)
 			return
 		}
-		attr = obj.Attribute(a)
+		attr = ref.Find(a)
 		if attr == nil {
 			eval.ReportError("%s type does not have an attribute named %#v", kind, a)
 			return

--- a/dsl/payload.go
+++ b/dsl/payload.go
@@ -1,8 +1,8 @@
 package dsl
 
 import (
-	"goa.design/goa/expr"
 	"goa.design/goa/eval"
+	"goa.design/goa/expr"
 )
 
 // Payload defines the data type of an method input. Payload also makes the

--- a/expr/attribute.go
+++ b/expr/attribute.go
@@ -260,19 +260,21 @@ func (a *AttributeExpr) Finalize() {
 	if ut, ok := a.Type.(UserType); ok {
 		ut.Attribute().Finalize()
 	}
-	for _, ref := range a.References {
-		ru, ok := ref.(UserType)
-		if !ok {
-			continue
+	if IsObject(a.Type) {
+		for _, ref := range a.References {
+			ru, ok := ref.(UserType)
+			if !ok {
+				continue
+			}
+			a.Inherit(ru.Attribute())
 		}
-		a.Inherit(ru.Attribute())
-	}
-	for _, base := range a.Bases {
-		ru, ok := base.(UserType)
-		if !ok {
-			continue
+		for _, base := range a.Bases {
+			ru, ok := base.(UserType)
+			if !ok {
+				continue
+			}
+			a.Merge(ru.Attribute())
 		}
-		a.Merge(ru.Attribute())
 	}
 }
 

--- a/expr/http_endpoint_test.go
+++ b/expr/http_endpoint_test.go
@@ -43,7 +43,7 @@ func TestEndpointValidation(t *testing.T) {
 		"endpoint-body-as-missed-payload-prop": {
 			DSL: testdata.EndpointBodyAsMissedPayloadProp,
 			Errors: []string{
-				"\"name\" is not found in result type in service \"Service\" HTTP endpoint \"Method\"",
+				"Request type does not have an attribute named \"name\" in service \"Service\" HTTP endpoint \"Method\"",
 			},
 		},
 		"endpoint-body-extend-payload": {

--- a/http/codegen/server_decode_test.go
+++ b/http/codegen/server_decode_test.go
@@ -141,6 +141,8 @@ func TestDecode(t *testing.T) {
 
 		{"body-primitive-array-user-validate", testdata.PayloadBodyPrimitiveArrayUserValidateDSL, testdata.PayloadBodyPrimitiveArrayUserValidateDecodeCode},
 		{"body-primitive-field-array-user", testdata.PayloadBodyPrimitiveFieldArrayUserDSL, testdata.PayloadBodyPrimitiveFieldArrayUserDecodeCode},
+		{"body-extend-primitive-field-array-user", testdata.PayloadExtendBodyPrimitiveFieldArrayUserDSL, testdata.PayloadBodyPrimitiveFieldArrayUserDecodeCode},
+		{"body-extend-primitive-field-string", testdata.PayloadExtendBodyPrimitiveFieldStringDSL, testdata.PayloadBodyPrimitiveFieldStringDecodeCode},
 		{"body-primitive-field-array-user-validate", testdata.PayloadBodyPrimitiveFieldArrayUserValidateDSL, testdata.PayloadBodyPrimitiveFieldArrayUserValidateDecodeCode},
 
 		{"body-query-object", testdata.PayloadBodyQueryObjectDSL, testdata.PayloadBodyQueryObjectDecodeCode},

--- a/http/codegen/testdata/payload_decode_functions.go
+++ b/http/codegen/testdata/payload_decode_functions.go
@@ -3620,6 +3620,29 @@ func DecodeMethodBodyPrimitiveArrayUserRequest(mux goahttp.Muxer, decoder func(*
 }
 `
 
+var PayloadBodyPrimitiveFieldStringDecodeCode = `// DecodeMethodBodyPrimitiveArrayUserRequest returns a decoder for requests
+// sent to the ServiceBodyPrimitiveArrayUser MethodBodyPrimitiveArrayUser
+// endpoint.
+func DecodeMethodBodyPrimitiveArrayUserRequest(mux goahttp.Muxer, decoder func(*http.Request) goahttp.Decoder) func(*http.Request) (interface{}, error) {
+	return func(r *http.Request) (interface{}, error) {
+		var (
+			body string
+			err  error
+		)
+		err = decoder(r).Decode(&body)
+		if err != nil {
+			if err == io.EOF {
+				return nil, goa.MissingPayloadError()
+			}
+			return nil, goa.DecodePayloadError(err.Error())
+		}
+		payload := NewMethodBodyPrimitiveArrayUserPayloadType(body)
+
+		return payload, nil
+	}
+}
+`
+
 var PayloadBodyPrimitiveFieldArrayUserValidateDecodeCode = `// DecodeMethodBodyPrimitiveArrayUserValidateRequest returns a decoder for
 // requests sent to the ServiceBodyPrimitiveArrayUserValidate
 // MethodBodyPrimitiveArrayUserValidate endpoint.

--- a/http/codegen/testdata/payload_dsls.go
+++ b/http/codegen/testdata/payload_dsls.go
@@ -1916,6 +1916,44 @@ var PayloadBodyPrimitiveFieldArrayUserDSL = func() {
 	})
 }
 
+var PayloadExtendBodyPrimitiveFieldArrayUserDSL = func() {
+	var Ext = Type("Ext", func() {
+		Attribute("b", String)
+	})
+	var PayloadType = Type("PayloadType", func() {
+		Extend(Ext)
+		Attribute("a", ArrayOf(String))
+	})
+	Service("ServiceBodyPrimitiveArrayUser", func() {
+		Method("MethodBodyPrimitiveArrayUser", func() {
+			Payload(PayloadType)
+			HTTP(func() {
+				POST("/")
+				Body("a")
+			})
+		})
+	})
+}
+
+var PayloadExtendBodyPrimitiveFieldStringDSL = func() {
+	var Ext = Type("Ext", func() {
+		Attribute("b", String)
+	})
+	var PayloadType = Type("PayloadType", func() {
+		Extend(Ext)
+		Attribute("a", ArrayOf(String))
+	})
+	Service("ServiceBodyPrimitiveArrayUser", func() {
+		Method("MethodBodyPrimitiveArrayUser", func() {
+			Payload(PayloadType)
+			HTTP(func() {
+				POST("/")
+				Body("b")
+			})
+		})
+	})
+}
+
 var PayloadBodyPrimitiveFieldArrayUserValidateDSL = func() {
 	var PayloadType = Type("PayloadType", func() {
 		Attribute("a", ArrayOf(String), func() {

